multinomial sampling should rely on binomial

YeungOnion opened the pull request:
addresses #183, but does not really fix it since binomial is naive sampling.

if we rely on binomial sampling then we can optimize once. This also updates the testing_boiler to work with Multinomial, notice how the tests are written to all use dynamically sized vectors and to rely on norms for approx absolute equality.

There is also a stoachastic test added, but I'm doubtful that I'm correct about both the test and sampling because it doesn't pass often enough. (Perhaps my choice of parameters in the test is not convergent enough to a normal distribution?) But a second pair of eyes on this would be helpful.

YeungOnion committed:
refactor(perf): Multinomial samples from Binomial

lint: fix lints

YeungOnion committed:
fix: correct sampling for multinomial, rely on Binomial

YeungOnion commented:
perhaps I need test statistics to assess if the samples are likely from the distribution...

YeungOnion commented on src/distribution/multinomial.rs:
```diff
@@ -166,14 +166,37 @@ where
+        // allocate result and write into it
+        let mut res = OVector::zeros_generic(self.p.shape_generic().0, nalgebra::U1);
+
+        for ind in 0..p_sorted_inds.len() - 1 {
```
thanks! I knew something was amiss, glad you caught it.

FreezyLemon commented:
> notice how the tests are written to all use dynamically sized vectors 

Hm... the static types (i.e. SMatrix, SVector) have more type safety and so I generally agree that the dynamically sized types need more testing. But I'm not sure about just testing the dynamic types; we might miss problems with specific nalgebra code paths or even incorrect trait bounds that would exclude the static types from being used in the `statrs` API altogether.

Looking at #165, this would also exclude basically all multivariate tests from `no_std` environments without an allocator.

Not 100% sure the existing boiler macro can cover everything... the `nalgebra`-based distributions are a fair bit different after all. But I think that e.g. `new_from_nalgebra` should be preferred over just `new` since it works for both types of Matrix/vector.

YeungOnion committed:
test: stochastic test for multinomial - need another pair of eyes

I'm still unsure if I'm sampling incorrectly or am using bad
verification technique, should rely on a common GoF test instead of my
homebrew statistics

YeungOnion commented:
I agree we should be testing our API size generics and that tests should work on no_std if we have it. I'll revert that.

Do you have any other ideas to verify that samples are correctly generated?

FreezyLemon commented:
> Do you have any other ideas to verify that samples are correctly generated?

Not really, to be honest. Except for very rudimentary testing (samples are in the correct domain/range of values), you'd probably need to implement proper statistical test methods..

YeungOnion commented:
I'll rely on other implementations of statistical tests and rv sampling to verify my implementation for now. Once we've got tests and verification for them, then I'll consider bootstrapping features.

YeungOnion committed:
test: mirror testing boiler where possible for multinomial testing

YeungOnion committed:
fix: remove unneeded trait bounds

also adds a comment demonstrating success of unwrap

YeungOnion commented:
Tests use both static and dynamic allocators, but aren't broken out into separate tests, but they aren't interleaved into the test suite, so shouldn't be a big lift to conditionally compile them for a `no_std` option.

YeungOnion commented:
Certainly not a thorough analysis, but it's safe to say the implementation is not likely correct.

```py
>>> stats.cramervonmises_2samp(scipy_mnom_gen(), data) # scipy vs data
CramerVonMisesResult(statistic=[4.46216824 4.365037   4.57871893 4.39759045 4.64458355], pvalue=[4.92114127e-11 7.65537633e-11 3.194799
98e-11 6.57013333e-11
 2.73471246e-11])
>>> stats.cramervonmises_2samp(scipy_mnom_gen(), scipy_mnom_gen()) # scipy vs scipy
CramerVonMisesResult(statistic=[0.1209659  0.05345861 0.15851277 0.10590302 0.13809556], pvalue=[0.49155904 0.8550833  0.36467217 0.556
74849 0.42800996])
```

tested with `n=1000, p=5*[0.2]` where both scipy and statrs generated 10k samples
